zdm/grid.py

```diff
@@ -514,125 +653,192 @@ def GenMCFRB(self, pwb=None, Emax_boost=0.0):
-                    + self.thresholds[j, iz2, iDM2] * kz2 * kDM2
-                )
+            # fraction of value through DM bin
+            fDM = dDM / (pDMc[iDM2] - pDMc[iDM1])
+            
+            # get the MC DM through interpolation
+            if fDM < 0.5:
+                kDM1 = 0.5 - fDM
+                kDM2 = 0.5 + fDM
+                kDM3 = 0.
+                iDM3 = 0    # dummy
+                # sets iDM3 to be safe at 0
+                MCDM = self.dmvals[iDM1] * kDM1 + self.dmvals[iDM2] * kDM2
+            elif iDM2 == self.dmvals.size-1:
+                kDM1 = 0.
+                kDM2 = 1. # for future use, not here
+                kDM3 = 0.
+                iDM1 = 0    # dummy
+                iDM3 = 0    # dummy
+                MCDM = self.dmvals[iDM2] + (fDM - 0.5)*self.dDM # upper DM bins
             else:
-                Eth = (
-                    self.thresholds[j, iz2, iDM1] * kDM1
-                    + self.thresholds[j, iz2, iDM2] * kDM2
-                )
-                Eth *= kz2 ** 2  # assume threshold goes as Eth~z^2 in the near Universe
+                kDM1 = 0.
+                kDM2 = 1.5-fDM
+                kDM3 = fDM - 0.5
+                iDM1 = 0    # dummy
+                
+                MCDM = self.dmvals[iDM3] * kDM3 + self.dmvals[iDM2] * kDM2
+            
+            #MCDM = self.dmvals[iDM1] * kDM1 + self.dmvals[iDM2] * kDM2
+            #if iz2 > 0:
+            #    Eth = (
+            #        self.thresholds[j, iz1, iDM1] * kz1 * kDM1
+            #        + self.thresholds[j, iz1, iDM2] * kz1 * kDM2
+            #        + self.thresholds[j, iz1, iDM3] * kz1 * kDM3
+            #        + self.thresholds[j, iz2, iDM1] * kz2 * kDM1
+            #        + self.thresholds[j, iz2, iDM2] * kz2 * kDM2
+            #        + self.thresholds[j, iz2, iDM3] * kz2 * kDM3
+            #        + self.thresholds[j, iz3, iDM1] * kz3 * kDM1
+            #        + self.thresholds[j, iz3, iDM2] * kz3 * kDM2
+            #        + self.thresholds[j, iz3, iDM3] * kz3 * kDM3
+            #    )
+            #else:
+            #    Eth = (
+            #        self.thresholds[j, iz2, iDM1] * kDM1
+            #        + self.thresholds[j, iz2, iDM2] * kDM2
+            #    )
+            #    Eth *= kz2 ** 2  # assume threshold goes as Eth~z^2 in the near Universe
         else:
             # interpolate linearly from 0 to the minimum value
-            kDM2 = r / pDMc[iDM2]
-            MCDM = self.dmvals[iDM2] * kDM2
-            if iz2 > 0:  # ignore effect of lowest DM bin on threshold
-                Eth = (
-                    self.thresholds[j, iz1, iDM2] * kz1
-                    + self.thresholds[j, iz2, iDM2] * kz2
-                )
-            else:
-                Eth = self.thresholds[j, iz2, iDM2] * kDM2
-                Eth *= kz2 ** 2  # assume threshold goes as Eth~z^2 in the near Universe
-
+            fDM = r / pDMc[iDM2]
```

Comment:
this also appears rather inefficient..
  but I'd have to look really carefully to see if there
  were obvious speed-ups